torch/functional.py

```diff
@@ -113,3 +118,91 @@ def btriunpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
         P = None
 
     return P, L, U
+
+def matmul(tensor1, tensor2, out=None):
+    """Matrix product of two tensors.
+
+    The behavior depends on the dimensionality of the tensors as follows:
+
+    - If both tensors are 1-dimensional, the dot product (scalar) is returned.
+    - If the first argument is 1-dimensional and the second argument is 2-dimensional,
+      a 1 is prepended to its dimension for the purpose of the matrix multiply.
+      After the matrix multiply, the prepended dimension is removed.
+    - If the first argument is 2-dimensional and the second argument is 1-dimensional,
+      the matrix-vector product is returned
+    - If both arguments are 2-dimensional, the matrix-matrix product is returned
+    - If both arguments are at least 2-dimensional and at least one argument is
+      N-dimensional (where N > 2), then a batched-matrix multiply is returned.  The
+      non-matrix dimensions are :ref:`broadcasted <broadcasting-semantics>` (and thus
+      must be broadcastable).  For example, if :attr:`tensor1` is a `j x 1 x n x m` Tensor
+      and :attr:`tensor2` is a `k x m x p` Tensor, :attr:`out` will be an `j x k x n x p` Tensor.
+
+    .. note::
+
+        The 1-dimensional dot product version of this function does not support an :attr:`out` parameter.
+
+    Arguments:
+        tensor1 (Tensor): First tensor to be multiplied
+        tensor2 (Tensor): Second tensor to be multiplied
+        out (Tensor, optional): Output tensor
+    """
+
+    dim_tensor1 = tensor1.dim()
+    try:
+        dim_tensor2 = tensor2.dim()
+    except AttributeError:  # not a tensor
+        return NotImplemented
```

Comment:
Ah I think it was there for a reason, but it no longer fits this place. `NotImplemented` is a special singleton that you can return from the operator overloads to indicate that you don't know what to do, and Python will raise a proper error message (`unsupported operand type(s) for +: 'int' and 'str'`) if both arguments return it. However in here, it will just return `NotImplemented` as if it was a regular value. So we should:
1. Add a type check in `__matmul__` that returns `NotImplemented` if other is not a tensor (we didn't have a check in first one, because it was `self`).
2. Remove the `try ... except ...` here